File: legacy_py/arbitrage/orderbook.py
"""Orderbook-based price estimation utilities."""

from __future__ import annotations

from typing import Iterable, Optional, Tuple

from .models import MarketQuote

Level = Tuple[float, float]


def _normalize_levels(levels: Iterable[Iterable[float]]) -> list[Level]:
    out: list[Level] = []
    for row in levels:
        try:
            price = float(row[0])
            amount = float(row[1])
        except (TypeError, ValueError, IndexError):
            continue
        if price > 0 and amount > 0:
            out.append((price, amount))
    return out


def estimate_fill_from_base(levels: Iterable[Iterable[float]], base_qty: float) -> Optional[float]:
    """Returns VWAP price for a base-quantity market fill."""
    if base_qty <= 0:
        return None

    remaining = base_qty
    cost = 0.0
    for price, amount in _normalize_levels(levels):
        take = min(amount, remaining)
        cost += price * take
        remaining -= take
        if remaining <= 1e-12:
            break

    if remaining > 1e-12:
        return None
    return cost / base_qty


def estimate_fill_from_quote(
    levels: Iterable[Iterable[float]],
    quote_amount: float,
) -> Optional[Tuple[float, float]]:
    """Returns (base_qty, VWAP price) for a quote-notional market fill."""
    if quote_amount <= 0:
        return None

    remaining = quote_amount
    base_qty = 0.0
    cost = 0.0
    for price, amount in _normalize_levels(levels):
        level_cost = price * amount
        if level_cost <= remaining:
            base_qty += amount
            cost += level_cost
            remaining -= level_cost
        else:
            partial = remaining / price
            base_qty += partial
            cost += remaining
            remaining = 0.0
            break

    if remaining > 1e-12 or base_qty <= 0:
        return None
    return base_qty, cost / base_qty


def quote_from_orderbook(orderbook: dict, base_qty: float) -> Optional[MarketQuote]:
    """Builds a MarketQuote (bid/ask VWAP) for a base quantity."""
    bids = orderbook.get("bids", []) if isinstance(orderbook, dict) else []
    asks = orderbook.get("asks", []) if isinstance(orderbook, dict) else []

    bid_price = estimate_fill_from_base(bids, base_qty)
    ask_price = estimate_fill_from_base(asks, base_qty)
    if not bid_price or not ask_price:
        return None
    return MarketQuote(bid=bid_price, ask=ask_price)

File: src/orderbook.ts
import { MarketQuote } from "./models";

export type Level = [number, number];

function normalizeLevels(levels: unknown): Level[] {
  if (!Array.isArray(levels)) return [];
  const out: Level[] = [];
  for (const row of levels) {
    if (!Array.isArray(row) || row.length < 2) continue;
    const price = Number(row[0]);
    const amount = Number(row[1]);
    if (Number.isFinite(price) && Number.isFinite(amount) && price > 0 && amount > 0) out.push([price, amount]);
  }
  return out;
}

function normalizeAsks(asks: unknown): Level[] {
  return normalizeLevels(asks).sort((a, b) => a[0] - b[0]);
}

function normalizeBids(bids: unknown): Level[] {
  return normalizeLevels(bids).sort((a, b) => b[0] - a[0]);
}

export function estimateFillFromBase(levels: unknown, baseQty: number): number | null {
  if (baseQty <= 0) return null;

  let remaining = baseQty;
  let cost = 0.0;
  for (const [price, amount] of normalizeLevels(levels)) {
    const take = Math.min(amount, remaining);
    cost += price * take;
    remaining -= take;
    if (remaining <= 1e-12) break;
  }

  if (remaining > 1e-12) return null;
  return cost / baseQty;
}

export function estimateFillFromQuote(levels: unknown, quoteAmount: number): { baseQty: number; price: number } | null {
  if (quoteAmount <= 0) return null;

  let remaining = quoteAmount;
  let baseQty = 0.0;
  let cost = 0.0;
  for (const [price, amount] of normalizeLevels(levels)) {
    const levelCost = price * amount;
    if (levelCost <= remaining) {
      baseQty += amount;
      cost += levelCost;
      remaining -= levelCost;
    } else {
      const partial = remaining / price;
      baseQty += partial;
      cost += remaining;
      remaining = 0.0;
      break;
    }
  }

  if (remaining > 1e-12 || baseQty <= 0) return null;
  return { baseQty, price: cost / baseQty };
}

export type FillAnalysis = {
  baseQty: number;
  quoteQty: number;
  vwapPrice: number;
  bestPrice: number;
  worstPrice: number;
  levels: number;
  impactPct: number;
};

export function analyzeBuyFillFromQuote(asks: unknown, quoteAmount: number): FillAnalysis | null {
  if (quoteAmount <= 0) return null;
  const levels = normalizeAsks(asks);
  const firstLevel = levels[0];
  if (!firstLevel) return null;

  const bestPrice = firstLevel[0];
  let remaining = quoteAmount;
  let baseQty = 0.0;
  let cost = 0.0;
  let worstPrice = bestPrice;
  let usedLevels = 0;

  for (const [price, amount] of levels) {
    if (remaining <= 1e-12) break;
    const levelCost = price * amount;
    usedLevels += 1;
    if (levelCost <= remaining) {
      baseQty += amount;
      cost += levelCost;
      remaining -= levelCost;
      worstPrice = price;
    } else {
      const partial = remaining / price;
      baseQty += partial;
      cost += remaining;
      remaining = 0.0;
      worstPrice = price;
      break;
    }
  }

  if (remaining > 1e-12 || baseQty <= 0) return null;
  const vwapPrice = cost / baseQty;
  const impactPct = ((worstPrice - bestPrice) / bestPrice) * 100.0;
  return { baseQty, quoteQty: quoteAmount, vwapPrice, bestPrice, worstPrice, levels: usedLevels, impactPct };
}

export function analyzeSellFillFromBase(bids: unknown, baseQty: number): FillAnalysis | null {
  if (baseQty <= 0) return null;
  const levels = normalizeBids(bids);
  const firstLevel = levels[0];
  if (!firstLevel) return null;

  const bestPrice = firstLevel[0];
  let remaining = baseQty;
  let proceeds = 0.0;
  let worstPrice = bestPrice;
  let usedLevels = 0;

  for (const [price, amount] of levels) {
    if (remaining <= 1e-12) break;
    const take = Math.min(amount, remaining);
    proceeds += price * take;
    remaining -= take;
    usedLevels += 1;
    worstPrice = price;
  }

  if (remaining > 1e-12) return null;
  const vwapPrice = proceeds / baseQty;
  const impactPct = ((bestPrice - worstPrice) / bestPrice) * 100.0;
  return { baseQty, quoteQty: proceeds, vwapPrice, bestPrice, worstPrice, levels: usedLevels, impactPct };
}

export function quoteFromOrderbook(orderbook: unknown, baseQty: number): MarketQuote | null {
  const bids = typeof orderbook === "object" && orderbook !== null ? (orderbook as any).bids : undefined;
  const asks = typeof orderbook === "object" && orderbook !== null ? (orderbook as any).asks : undefined;

  const bidPrice = estimateFillFromBase(bids, baseQty);
  const askPrice = estimateFillFromBase(asks, baseQty);
  if (!bidPrice || !askPrice) return null;
  return { bid: bidPrice, ask: askPrice };
}
